File: pkg/runtime/runtimes/go/pulserpc/contract.go
package pulserpc

import (
	"encoding/json"
	"fmt"
	"reflect"
)

// FunctionDef represents a function/method definition from the IDL
type FunctionDef map[string]interface{}

// Interface represents an interface from the IDL
type Interface struct {
	Name      string
	Functions map[string]FunctionDef
}

// Contract represents a parsed IDL contract
type Contract struct {
	idlParsed  interface{}
	Interfaces map[string]*Interface
	Structs    StructMap
	Enums      EnumMap
	Meta       map[string]interface{}
}

// NewContract creates a new Contract from parsed IDL data
// Supports both PulseRPC format (dict with interfaces, structs, enums keys)
// and Barrister format (list of items with type field)
func NewContract(idlParsed interface{}) *Contract {
	c := &Contract{
		idlParsed:  idlParsed,
		Interfaces: make(map[string]*Interface),
		Structs:    make(StructMap),
		Enums:      make(EnumMap),
		Meta:       make(map[string]interface{}),
	}

	// Handle both barrister format (list) and PulseRPC format (dict)
	if dict, ok := idlParsed.(map[string]interface{}); ok {
		// PulseRPC format - dict with interfaces, structs, enums keys

		// Parse interfaces
		if interfaces, ok := dict["interfaces"].([]interface{}); ok {
			for _, ifaceData := range interfaces {
				if ifaceDict, ok := ifaceData.(map[string]interface{}); ok {
					iface := &Interface{
						Name:      ifaceDict["name"].(string),
						Functions: make(map[string]FunctionDef),
					}
					if methods, ok := ifaceDict["methods"].([]interface{}); ok {
						for _, method := range methods {
							if methodDict, ok := method.(map[string]interface{}); ok {
								funcName := methodDict["name"].(string)
								iface.Functions[funcName] = methodDict
							}
						}
					}
					c.Interfaces[iface.Name] = iface
				}
			}
		}

		// Parse structs
		if structs, ok := dict["structs"].([]interface{}); ok {
			for _, structData := range structs {
				if structDict, ok := structData.(map[string]interface{}); ok {
					name := structDict["name"].(string)
					c.Structs[name] = structDict
				}
			}
		}

		// Parse enums
		if enums, ok := dict["enums"].([]interface{}); ok {
			for _, enumData := range enums {
				if enumDict, ok := enumData.(map[string]interface{}); ok {
					name := enumDict["name"].(string)
					c.Enums[name] = enumDict
				}
			}
		}
	} else if list, ok := idlParsed.([]interface{}); ok {
		// Barrister format - list of items with type field
		for _, item := range list {
			if itemDict, ok := item.(map[string]interface{}); ok {
				itemType, _ := itemDict["type"].(string)
				switch itemType {
				case "struct":
					if name, ok := itemDict["name"].(string); ok {
						c.Structs[name] = itemDict
					}
				case "enum":
					if name, ok := itemDict["name"].(string); ok {
						c.Enums[name] = itemDict
					}
				case "interface":
					iface := &Interface{
						Name:      itemDict["name"].(string),
						Functions: make(map[string]FunctionDef),
					}
					if methods, ok := itemDict["methods"].([]interface{}); ok {
						for _, method := range methods {
							if methodDict, ok := method.(map[string]interface{}); ok {
								funcName := methodDict["name"].(string)
								iface.Functions[funcName] = methodDict
							}
						}
					}
					c.Interfaces[iface.Name] = iface
				case "meta":
					// Copy metadata
					for key, value := range itemDict {
						if key != "type" {
							c.Meta[key] = value
						}
					}
				}
			}
		}
	}

	return c
}

// HasInterface checks if an interface exists in the contract
func (c *Contract) HasInterface(ifaceName string) bool {
	_, ok := c.Interfaces[ifaceName]
	return ok
}

// GetInterface returns an interface by name, or nil if not found
func (c *Contract) GetInterface(ifaceName string) *Interface {
	iface, ok := c.Interfaces[ifaceName]
	if !ok {
		return nil
	}
	return iface
}

// GetFunction returns a function definition from an interface
func (i *Interface) GetFunction(funcName string) FunctionDef {
	if fn, ok := i.Functions[funcName]; ok {
		return fn
	}
	return nil
}

// ValidateRequest validates request parameters against the IDL
func (c *Contract) ValidateRequest(ifaceName string, funcName string, params []interface{}) error {
	iface := c.GetInterface(ifaceName)
	if iface == nil {
		return fmt.Errorf("unknown interface: '%s'", ifaceName)
	}

	fn := iface.GetFunction(funcName)
	if fn == nil {
		return fmt.Errorf("%s: unknown function: '%s'", ifaceName, funcName)
	}

	// Get parameter definitions
	paramDefs, ok := fn["parameters"].([]interface{})
	if !ok {
		paramDefs = []interface{}{}
	}

	// Check parameter count
	if len(params) != len(paramDefs) {
		return fmt.Errorf("function '%s.%s' expects %d param(s), %d given",
			ifaceName, funcName, len(paramDefs), len(params))
	}

	// Validate each parameter
	for i, paramValue := range params {
		paramDef, ok := paramDefs[i].(map[string]interface{})
		if !ok {
			continue
		}

		paramName, _ := paramDef["name"].(string)
		paramType, ok := paramDef["type"].(map[string]interface{})
		if !ok {
			continue
		}

		isOptional := false
		if opt, ok := paramDef["optional"].(bool); ok {
			isOptional = opt
		}

		if err := ValidateType(paramValue, paramType, c.Structs, c.Enums, isOptional); err != nil {
			return fmt.Errorf("function '%s.%s' invalid param '%s': %w",
				ifaceName, funcName, paramName, err)
		}
	}

	return nil
}

// ValidateResponse validates response result against the IDL
func (c *Contract) ValidateResponse(ifaceName string, funcName string, result interface{}) error {
	iface := c.GetInterface(ifaceName)
	if iface == nil {
		return fmt.Errorf("unknown interface: '%s'", ifaceName)
	}

	fn := iface.GetFunction(funcName)
	if fn == nil {
		return fmt.Errorf("%s: unknown function: '%s'", ifaceName, funcName)
	}

	// Check if function has a return type
	returnType, ok := fn["returnType"].(map[string]interface{})
	if !ok {
		// Function returns void/None
		if result != nil {
			return fmt.Errorf("function '%s.%s' invalid response: expected nil, got %v",
				ifaceName, funcName, result)
		}
		return nil
	}

	// Convert result to map if it's a struct (for proper validation)
	result, err := convertToMap(result)
	if err != nil {
		return fmt.Errorf("function '%s.%s' invalid response: %w",
			ifaceName, funcName, err)
	}

	// Validate return type
	isOptional := false
	if opt, ok := fn["returnOptional"].(bool); ok {
		isOptional = opt
	}

	if err := ValidateType(result, returnType, c.Structs, c.Enums, isOptional); err != nil {
		return fmt.Errorf("function '%s.%s' invalid response: %w",
			ifaceName, funcName, err)
	}

	return nil
}

// convertToMap converts a Go struct to map[string]interface{} for validation
// If the value is already a map or primitive, it returns it as-is
func convertToMap(value interface{}) (interface{}, error) {
	if value == nil {
		return nil, nil
	}

	// If it's already a map or primitive, return as-is
	switch value.(type) {
	case map[string]interface{}, string, float64, int, bool, []interface{}:
		return value, nil
	}

	// For pointers, unwrap and try again
	if rv := reflect.ValueOf(value); rv.Kind() == reflect.Ptr {
		if rv.IsNil() {
			return nil, nil
		}
		return convertToMap(rv.Elem().Interface())
	}

	// For slices/arrays, convert each element
	if rv := reflect.ValueOf(value); rv.Kind() == reflect.Slice || rv.Kind() == reflect.Array {
		result := make([]interface{}, rv.Len())
		for i := 0; i < rv.Len(); i++ {
			elem, err := convertToMap(rv.Index(i).Interface())
			if err != nil {
				return nil, err
			}
			result[i] = elem
		}
		return result, nil
	}

	// For structs, convert to map via JSON
	jsonBytes, err := json.Marshal(value)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal struct to JSON: %w", err)
	}

	var result map[string]interface{}
	if err := json.Unmarshal(jsonBytes, &result); err != nil {
		return nil, fmt.Errorf("failed to unmarshal JSON to map: %w", err)
	}

	return result, nil
}

File: pkg/runtime/runtimes/csharp/PulseRPC/Contract.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace PulseRPC
{
    /// <summary>
    /// Represents a function/method definition from the IDL
    /// </summary>
    public class FunctionDef : Dictionary<string, object>
    {
    }

    /// <summary>
    /// Represents an interface from the IDL
    /// </summary>
    public class Interface
    {
        public string Name { get; set; } = "";
        public Dictionary<string, FunctionDef> Functions { get; set; } = new Dictionary<string, FunctionDef>();

        /// <summary>
        /// Get a function definition by name
        /// </summary>
        public FunctionDef? GetFunction(string funcName)
        {
            return Functions.TryGetValue(funcName, out var fn) ? fn : null;
        }
    }

    /// <summary>
    /// Represents a parsed IDL contract
    /// </summary>
    public class Contract
    {
        public object IdlParsed { get; set; } = null!;
        public Dictionary<string, Interface> Interfaces { get; set; } = new Dictionary<string, Interface>();
        public Dictionary<string, Dictionary<string, object>> Structs { get; set; } = new Dictionary<string, Dictionary<string, object>>();
        public Dictionary<string, Dictionary<string, object>> Enums { get; set; } = new Dictionary<string, Dictionary<string, object>>();
        public Dictionary<string, object> Meta { get; set; } = new Dictionary<string, object>();

        /// <summary>
        /// Creates a new Contract from parsed IDL data.
        /// Supports both PulseRPC format (dict with interfaces, structs, enums keys)
        /// and Barrister format (list of items with type field)
        /// </summary>
        public Contract(object idlParsed)
        {
            IdlParsed = idlParsed;

            // Handle both barrister format (list) and PulseRPC format (dict)
            if (idlParsed is Dictionary<string, object> dict)
            {
                // PulseRPC format - dict with interfaces, structs, enums keys

                // Parse interfaces
                if (dict.TryGetValue("interfaces", out var interfacesObj) && interfacesObj is System.Collections.IList interfacesList)
                {
                    foreach (var ifaceData in interfacesList)
                    {
                        if (ifaceData is Dictionary<string, object> ifaceDict && ifaceDict.TryGetValue("name", out var nameObj))
                        {
                            var iface = new Interface
                            {
                                Name = nameObj?.ToString() ?? ""
                            };

                            if (ifaceDict.TryGetValue("methods", out var methodsObj) && methodsObj is System.Collections.IList methodsList)
                            {
                                foreach (var method in methodsList)
                                {
                                    if (method is Dictionary<string, object> methodDict && methodDict.TryGetValue("name", out var funcNameObj))
                                    {
                                        var funcName = funcNameObj?.ToString() ?? "";
                                        var funcDef = new FunctionDef();
foreach (var kvp in methodDict)
{
    funcDef[kvp.Key] = kvp.Value;
}
iface.Functions[funcName] = funcDef;
                                    }
                                }
                            }
                            Interfaces[iface.Name] = iface;
                        }
                    }
                }

                // Parse structs
                if (dict.TryGetValue("structs", out var structsObj) && structsObj is System.Collections.IList structsList)
                {
                    foreach (var structData in structsList)
                    {
                        if (structData is Dictionary<string, object> structDict && structDict.TryGetValue("name", out var structNameObj))
                        {
                            var name = structNameObj?.ToString() ?? "";
                            Structs[name] = structDict;
                        }
                    }
                }

                // Parse enums
                if (dict.TryGetValue("enums", out var enumsObj) && enumsObj is System.Collections.IList enumsList)
                {
                    foreach (var enumData in enumsList)
                    {
                        if (enumData is Dictionary<string, object> enumDict && enumDict.TryGetValue("name", out var enumNameObj))
                        {
                            var name = enumNameObj?.ToString() ?? "";
                            Enums[name] = enumDict;
                        }
                    }
                }
            }
            else if (idlParsed is System.Collections.IList list)
            {
                // Barrister format - list of items with type field
                foreach (var item in list)
                {
                    if (item is Dictionary<string, object> itemDict && itemDict.TryGetValue("type", out var typeObj))
                    {
                        var itemType = typeObj?.ToString();
                        switch (itemType)
                        {
                            case "struct":
                                if (itemDict.TryGetValue("name", out var structNameObj))
                                {
                                    var name = structNameObj?.ToString() ?? "";
                                    Structs[name] = itemDict;
                                }
                                break;
                            case "enum":
                                if (itemDict.TryGetValue("name", out var enumNameObj))
                                {
                                    var name = enumNameObj?.ToString() ?? "";
                                    Enums[name] = itemDict;
                                }
                                break;
                            case "interface":
                                var iface = new Interface();
                                if (itemDict.TryGetValue("name", out var ifaceNameObj))
                                {
                                    iface.Name = ifaceNameObj?.ToString() ?? "";
                                }
                                if (itemDict.TryGetValue("methods", out var methodsObj) && methodsObj is System.Collections.IList methodsList)
                                {
                                    foreach (var method in methodsList)
                                    {
                                        if (method is Dictionary<string, object> methodDict && methodDict.TryGetValue("name", out var funcNameObj))
                                        {
                                            var funcName = funcNameObj?.ToString() ?? "";
                                            var funcDef = new FunctionDef();
foreach (var kvp in methodDict)
{
    funcDef[kvp.Key] = kvp.Value;
}
iface.Functions[funcName] = funcDef;
                                        }
                                    }
                                }
                                Interfaces[iface.Name] = iface;
                                break;
                            case "meta":
                                // Copy metadata
                                foreach (var kvp in itemDict)
                                {
                                    if (kvp.Key != "type")
                                    {
                                        Meta[kvp.Key] = kvp.Value;
                                    }
                                }
                                break;
                        }
                    }
                }
            }
        }

        /// <summary>
        /// Check if an interface exists in the contract
        /// </summary>
        public bool HasInterface(string ifaceName)
        {
            return Interfaces.ContainsKey(ifaceName);
        }

        /// <summary>
        /// Get an interface by name
        /// </summary>
        public Interface? GetInterface(string ifaceName)
        {
            return Interfaces.TryGetValue(ifaceName, out var iface) ? iface : null;
        }

        /// <summary>
        /// Validate request parameters against the IDL
        /// </summary>
        public void ValidateRequest(string ifaceName, string funcName, List<object?> @params)
        {
            var iface = GetInterface(ifaceName);
            if (iface == null)
            {
                throw new ArgumentException($"Unknown interface: '{ifaceName}'");
            }

            var fn = iface.GetFunction(funcName);
            if (fn == null)
            {
                throw new ArgumentException($"{ifaceName}: Unknown function: '{funcName}'");
            }

            // Get parameter definitions
            var paramDefs = fn.TryGetValue("parameters", out var paramDefsObj) && paramDefsObj is System.Collections.IList paramList
                ? paramList
                : new System.Collections.ArrayList();

            // Check parameter count
            if (@params.Count != paramDefs.Count)
            {
                throw new ArgumentException($"Function '{ifaceName}.{funcName}' expects {paramDefs.Count} param(s), {@params.Count} given");
            }

            // Validate each parameter
            for (int i = 0; i < @params.Count; i++)
            {
                var paramValue = @params[i];
                var paramDef = paramDefs[i] as Dictionary<string, object>;
                if (paramDef == null) continue;

                var paramName = paramDef.TryGetValue("name", out var pn) ? pn?.ToString() ?? "" : "";
                var paramType = paramDef.TryGetValue("type", out var pt) && pt is Dictionary<string, object> ptDict
                    ? ptDict
                    : new Dictionary<string, object>();

                var isOptional = paramDef.TryGetValue("optional", out var optObj) && optObj is bool opt && opt;

                try
                {
                    Validation.ValidateType(paramValue, paramType, Structs, Enums, isOptional);
                }
                catch (Exception e)
                {
                    throw new ArgumentException($"Function '{ifaceName}.{funcName}' invalid param '{paramName}': {e.Message}", e);
                }
            }
        }

        /// <summary>
        /// Validate response result against the IDL
        /// </summary>
        public void ValidateResponse(string ifaceName, string funcName, object? result)
        {
            var iface = GetInterface(ifaceName);
            if (iface == null)
            {
                throw new ArgumentException($"Unknown interface: '{ifaceName}'");
            }

            var fn = iface.GetFunction(funcName);
            if (fn == null)
            {
                throw new ArgumentException($"{ifaceName}: Unknown function: '{funcName}'");
            }

            // Check if function has a return type
            if (!fn.TryGetValue("returnType", out var returnTypeObj) || returnTypeObj is not Dictionary<string, object> returnType)
            {
                // Function returns void/None
                if (result != null)
                {
                    throw new ArgumentException($"Function '{ifaceName}.{funcName}' invalid response: expected null, got {result}");
                }
                return;
            }

            // Validate return type
            var isOptional = fn.TryGetValue("returnOptional", out var retOptObj) && retOptObj is bool retOpt && retOpt;

            try
            {
                Validation.ValidateType(result, returnType, Structs, Enums, isOptional);
            }
            catch (Exception e)
            {
                throw new ArgumentException($"Function '{ifaceName}.{funcName}' invalid response: {e.Message}", e);
            }
        }
    }
}
